### DISCUSSION TITLE
using 2nd core

### DISCUSSION THREAD

**Discussion by danilkorotkov** [HomeSpan/HomeSpan#262] [Ideas]
a little improvement in setup function makes homespan more fast & reliable (for 2 core esp32)
tested with my recuperator sketch

in the head 
`TaskHandle_t h_HK_poll;`

setup:
```
  xTaskCreatePinnedToCore(
                    HK_poll,    /* Task function. */
                    "HK_poll",  /* name of task. */
                    10000,       /* Stack size of task */
                    NULL,        /* parameter of the task */
                    1,           /* priority of the task */
                    &h_HK_poll, /* Task handle to keep track of created task */
                    0);          /* pin task to core 0 */  
  
  delay(1000);
```

the task
```
void HK_poll(void * pvParameters){

  for (;;){
    homeSpan.poll();
  }   //loop
}     //task
```

**Comment by HomeSpan** [comment]
Thanks - I've not used native IDF task handling but this looks promising.  Do you know if it works with the single-core S2 and C3 chips as well (it may not provide any benefits, but does the logic still work?)

**Reply by danilkorotkov** [reply]
Dual mcu way.  Tuya devices…

**Reply by danilkorotkov** [reply]
Unfortunately, reliable stepper motor is the task for standalone hardware/driver. 20 years ago I made a dual 8051 mcu board (asm + notepad!). 1st for stable motor timing, 2nd for serial port to pc. :)

**Comment by Xorlent** [comment]
How are you also pinning Wi-Fi tasks to a specific core?  My understanding is this is only possible if you are using ESP-IDF.  Given that assumption, I wonder if there is any benefit to changing the default core.

Update: I see now, it looks like Wi-Fi tasks are, by default, pinned to core 0.

**Reply by danilkorotkov** [reply]
This code also works with arduino ide. 
Default arduino core is 1. 
Core 0 is free of any task by default

**Comment by Olivierbkk** [comment]
Hi,

I am trying this but it doesn't work, the wifi led on the board is off also. here is the code:

```C++
TaskHandle_t h_HK_poll; // to run on homespan core 0

#include "HomeSpan.h"
#include "DEV_Curtains.h"

void setup() {

 // to run on homespan core 0
  xTaskCreatePinnedToCore(
                    HK_poll,    /* Task function. */
                    "HK_poll",  /* name of task. */
                    10000,       /* Stack size of task */
                    NULL,        /* parameter of the task */
                    1,           /* priority of the task */
                    &h_HK_poll, /* Task handle to keep track of created task */
                    0);          /* pin task to core 0 */  
  
  delay(1000);

  Serial.begin(115200);

  homeSpan.begin(Category::WindowCoverings, "Curtains");

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Name("Corner curtain");
      new Characteristic::Identify();
    new DEV_Curtain();
}

/*  // commented to run on homespan core 0 - see replacement below
void loop() {
  homeSpan.poll();
}*/

void HK_poll(void * pvParameters){
  for (;;){
    homeSpan.poll();
  }   //loop
}     //task
```

the serial monitor returns this:

```
Welcome to HomeSpan!
Apple HomeKit for the Espressif ESP-32 WROOM and Arduino IDE

Please ensure serial monitor is set to transmit <newlines>

Message Logs:     Level 0
Status LED:       Pin -  *** WARNING: Status LED Pin is UNDEFINED
Device Control:   Pin -  *** WARNING: Device Control Pin is UNDEFINED
Sketch Version:   n/a
HomeSpan Version: 1.7.1
Arduino-ESP Ver.: 2.0.6
ESP-IDF Version:  4.4.3
ESP32 Chip:       ESP32-D0WDQ5 Rev 3 dual-core 4MB Flash
ESP32 Board:      esp32da
PWM Resources:    16 channels, 8 timers, max 20-bit duty resolution
Sodium Version:   1.0.12-idf  Lib 9.4
MbedTLS Version:  mbed TLS 2.28.1
Sketch Compiled:  Mar 22 2023 08:34:28
Partition:        app0
MAC Address:      B8:B6:1A:66:C5:AD

Device Name:      Curtains
```

Can you tell me what I'm doing wrong?

**Reply by danilkorotkov** [reply]
do not comment out all loop(), only exclude .poll();

init first all homekit through homespan then create the task 
or you can do

```
void HK_poll(void * pvParameters){
  homeSpan.begin(Category::WindowCoverings, "Curtains");

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Name("Corner curtain");
      new Characteristic::Identify();
    new DEV_Curtain();
  for (;;){
    homeSpan.poll();
  }   //loop
}     //task

```
but it is better to init homekit separetly on the default core not inside the task

**Comment by Olivierbkk** [comment]
i updated the code like this:
```C++
TaskHandle_t h_HK_poll; // to run on homespan core 0

#include "HomeSpan.h"
#include "DEV_Curtains.h"

void setup() {

  xTaskCreatePinnedToCore(
                    HK_poll,    /* Task function. */
                    "HK_poll",  /* name of task. */
                    10000,       /* Stack size of task */
                    NULL,        /* parameter of the task */
                    1,           /* priority of the task */
                    &h_HK_poll, /* Task handle to keep track of created task */
                    0);          /* pin task to core 0 */  
  
  delay(1000);

  Serial.begin(115200);
}

void HK_poll(void * pvParameters){
  homeSpan.begin(Category::WindowCoverings, "Curtains");

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Name("Corner curtain");
      new Characteristic::Identify();
    new DEV_Curtain();
  
  for (;;){
    homeSpan.poll();
  }   //loop
}     //task

void loop() {
  homeSpan;
}
```

now the wifi led is blinking as usual, I can enter the wifi credentials but I cant' pair to homekit

**Reply by danilkorotkov** [reply]
void loop() {
  //homeSpan;
}
void setup() {
Serial.begin(115200);…

**Reply by danilkorotkov** [reply]
set up your own pair code in the serial monitor

**Reply by Olivierbkk** [reply]
I have absolutely no clue of what I am doing...

here is the code now, I still can't pair even after setting my own pairing code as you suggested

```C++
TaskHandle_t h_HK_poll; // to run on homespan core 0

#include "HomeSpan.h"
#include "DEV_Curtains.h"

void loop() {
//homeSpan;
}

void setup() {
  Serial.begin(115200);

  xTaskCreatePinnedToCore(
                    HK_poll,    /* Task function. */
                    "HK_poll",  /* name of task. */
                    10000,       /* Stack size of task */
                    NULL,        /* parameter of the task */
                    1,           /* priority of the task */
                    &h_HK_poll, /* Task handle to keep track of created task */
                    0);          /* pin task to core 0 */  
  
  delay(1000);
}

void HK_poll(void * pvParameters){
  homeSpan.begin(Category::WindowCoverings, "Curtains");

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Name("Corner curtain");
      new Characteristic::Identify();
    new DEV_Curtain();
  
  for (;;){
    homeSpan.poll();
  }   //loop
}     //task```

**Reply by danilkorotkov** [reply]
the pairing progress log from the serial monitor?


```
TaskHandle_t h_HK_poll; // to run on homespan core 0

#include "HomeSpan.h"
#include "DEV_Curtains.h"

void HK_poll(void * pvParameters){
  for (;;){
    homeSpan.poll();
  }   //loop
} 

void setup() {
  Serial.begin(115200);
  homeSpan.begin(Category::WindowCoverings, "Curtains");

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Name("Corner curtain");
      new Characteristic::Identify();
    new DEV_Curtain();

  xTaskCreatePinnedToCore(
                    HK_poll,    /* Task function. */
                    "HK_poll",  /* name of task. */
                    10000,       /* Stack size of task */
                    NULL,        /* parameter of the task */
                    1,           /* priority of the task */
                    &h_HK_poll, /* Task handle to keep track of created task */
                    0);          /* pin task to core 0 */  
  
  delay(1000);
}

void loop() {

} 
```

**Reply by Olivierbkk** [reply]
Message Logs:     Level 0
Status LED:       Pin -  *** WARNING: Status LED Pin is UNDEFINED
Device Control:   Pin -  *** WARNING: Device Control Pin is UNDEFINED
Sketch Version:   n/a
HomeSpan Version: 1.7.1
Arduino-ESP Ver.: 2.0.6
ESP-IDF Version:  4.4.3
ESP32 Chip:       ESP32-D0WDQ5 Rev 3 dual-core 4MB Flash
ESP32 Board:      esp32da
PWM Resources:    16 channels, 8 timers, max 20-bit duty resolution
Sodium Version:   1.0.12-idf  Lib 9.4
MbedTLS Version:  mbed TLS 2.28.1
Sketch Compiled:  Mar 23 2023 16:45:26
Partition:        app0
MAC Address:      B8:D6:1A:68:C5:AC

Device Name:      Curtains

initial, currentPosition/current = 0/0 targetPosition/target = 0/0

*** HomeSpan Info ***

➤ Accessory:  AID=1
   ➟ Service AccessoryInformation:  IID=1, UUID="3E"
      ⇨ Characteristic Name("Corner curtain"):  IID=2, UUID="23", Perms=PR
      ⇨ Characteristic Identify(0):  IID=3, UUID="14", Perms=PW
   ➟ Service WindowCovering:  IID=4, UUID="8C"
      ⇨ Characteristic CurrentPosition(0):  IID=5, UUID="6D", Perms=PR+EV, Custom-Range=[0,100,1] (nvs)
      ⇨ Characteristic TargetPosition(0):  IID=6, UUID="7C", Perms=PR+PW+EV, Custom-Range=[0,100,1] (nvs)
      ⇨ Characteristic PositionState(2):  IID=7, UUID="72", Perms=PR+EV, Range=[0,2]
      ⇨ Characteristic ObstructionDetected(0):  IID=8, UUID="24", Perms=PR+EV

Configured as Bridge: NO

Database Validation:  Warnings=0, Errors=0

Service                             UUID         AID  IID  Update  Loop  Button  Linked Services
------------------------------  --------  ----------  ---  ------  ----  ------  ---------------
AccessoryInformation                  3E           1    1      NO    NO      NO  -
WindowCovering                        8C           1    4     YES   YES      NO  -

*** End Info ***

Generating SRP verification data for default Setup Code: 466-37-726
Setup Payload for Optional QR Code: X-HM://00DYFDHUM    

Generating new random Accessory ID and Long-Term Ed25519 Signature Keys...
Initializing storage for Paired Controllers data...

Accessory ID:      C7:E6:1A:16:C5:10                               LTPK: BFD9BFD98B508ADE7FF1287F1A3860592883164542FC9F4E01DD88984EFBE99E
No Paired Controllers
Resetting Database Hash...

Accessory configuration has changed.  Updating configuration number to 1

**Comment by Olivierbkk** [comment]
ok it works now, it needs to have the "erase all flash before sketch upload" enabled

thanks for your patience...

**Comment by HomeSpan** [comment]
Note that homeSpan.autoPoll() allows you to move all polling onto a separate thread, which will also be on the second core for a dual core device.  

This allows you to add code to the main loop() function that runs in parallel and is not blocked by homeSpan polling. 

See the API for details.

**Reply by danilkorotkov** [reply]
`void autoPoll(uint32_t stackSize=CONFIG_ARDUINO_LOOP_STACK_SIZE){xTaskCreateUniversal([](void *parms){for(;;)homeSpan.pollTask();}, "pollTask", stackSize, NULL, 1, &pollTaskHandle, 0);}     // start pollTask()
`
Is xTaskCreateUniversal Arduino IDF only?

**Reply by HomeSpan** [reply]
It's defined in the Arduino-ESP32 *esp32-hal.h* file.

**Reply by danilkorotkov** [reply]
I didn’t find it in the original espressif freertos api

**Comment by Olivierbkk** [comment]
yes! that works too and seems to be even better for my application where I send pulses to a step motor. I get proper status now in the home app.

Thanks
